### ISSUE TITLE
Define recommended ruleset for v4.0.0

### ISSUE THREAD

**Issue by juanpcapurro** [solhint-community/solhint-community#89]
This issue reviews all the currently existing rules and tries to be a place
where we can discuss what's a sane default for the recommended ruleset. 

## TL;DR
Please comment here if you disagree with any of the following

- [x] making all the `.*-mixedCase` rules not recommended (and perhaps even deprecated), and create a `\0-camelCase` equivalent (although adding the possibility of having a leading underscore for private/non-state vars), and make that the recommended covered by #108 and #114 , although I kept the `mixedCase` naming because that's how it's referred to in the style guide
- [x] `contract-name-mixedcase` in particular should be replaced by
`contract-name-capwords` and enforce it starts with a capital letter covered by  #108
- [x] adding `interface-starts-with-i` from defi-wonderland/solhint-plugin to
the main rules as recommended -- done in #107
- [ ] making the following rules recommended, which currently aren't:
    ~- [ ] `code-complexity`~
    - [x] `custom-errors` #106
    - [x] `explicit-types` #106
    - [x] `one-contract-per-file` -- #115
    - [x] `modifier-name-mixedcase` -- #115
    ~- [ ] `named-parameters-mapping`~
    ~- [ ] `named-parameters-function`~
    - [ ] `named-return-values` added in #106 , but a sane default would be to allow only one anonymous return value, and require an explicit name if there are more than one, see #116
- [x] removing the `func-order` rule -- done in #103
- [x] removing `no-unused-import,` making it a special case of `no-unused-var` -- done in #22
- [x] making `not-rely-on-time` NOT recommended, which currently is #106

# Comprehensive review of existing rules:

### code-complexity.js
should be recommended: yes, on warning (not very convinced tho)
currently it's not recommended, and default severity is error

### custom-errors.js
should be recommended: yes, on error
currently not recommended

### explicit-types.js
should be recommended: yes, on warning
this would be the same behaviour as `forge fmt`'s default settings

### function-max-lines.js
should be recommended: no (currently isn't)
most real world projects have a fair deal of inherent inline complexity with no obvious way to get rid of

### max-line-length.js
should be recommended: no (currently isn't)
while I personally like it a lot, it's hard to sell on existing codebases and ideally is covered by prettier.

### max-states-count.js
should be recommended: yes (weak)
I don't like it that much (tough sell on existing codebases) but considering it's already recommended, I think it can stay

### no-console.js
should be recommended: yes
I imagine it being useful on a default CI. this would *require* however, that we implement #84 to disable it easily on the `script/` directory, for example

### no-empty-blocks.js
should be recommended: yes
already recommended, cant imagine it being that annoying

### no-global-import.js
should be recommended: yes (already is)
not a tough sell on existing codebases, people now take care to not bring things into scope willy-nilly

### no-unused-import.js
should be merged into no-unused-vars below.  (#23)

### no-unused-vars.js
should be recommended: yes (already is)

### one-contract-per-file.js
should be recommended: yes
only makes sense to disable it on tests implementing some kind of branching tree technique, or for small mocks perhaps. That should be taken care of by #84

### payable-fallback.js
should be recommended: yes (already is)
its docs can be improved though

### reason-string.js
should be recommended: yes (already is)

### constructor-syntax.js
should be recommended: yes (currently isn't)

### comprehensive-interface.js
should be recommended: no
I'd like to make it recommended but would obviously give lots of errors on foundry test contracts.
docs could be improved

### quotes.js
should be recommended: yes (already is)
should perhaps get a fixer? I think that's already covered by prettier but I guess that can't hurt. Good first issue.

### const-name-snakecase.js
should be recommended: yes (already is)

### contract-name-camelcase.js
should be recommended: yes (already is)
ACHSUALLY we should enforce the contract name starts with a capital letter as
well as it being camelCase

### event-name-camelcase.js
should be recommended: yes (already is)

### foundry-test-functions.js
should be recommended: no
depends on #84 to have good DX. Check the actual grammar of foundry test functions.
same as comprehensive-interface, can't be recommended since it'd report lots of errors on all non-test files

### func-name-mixedcase.js
should be recommended: yes (already is)
reports errors on foundry test functions, but in that case users'll have foundry-test-functions enabled manually anyway.
it's sad we have both mixedCase( `/[_]*[a-z$]+[a-zA-Z0-9$]*[_]?/` ) and camelCase (`/[A-Z$]+[a-zA-Z0-9$]*/`) as two subtly different things. I'd love to get rid of that distinction.

### func-param-name-mixedcase.js
should be recommended: yes
I believe it should actually be camelCase, to disallow snake_case

### immutable-name-snakecase.js
should be recommended:  yes

### modifier-name-mixedcase.js
should be recommended:  yes
I believe it should actually be camelCase, to disallow snake_case

### named-parameters-function.js
should be recommended: no (very weak, please contribute your opinion)
I believe there is much to gain from not having to remember/look up
the semantic meaning of every positional argument when _reading_ code, even if it's at the cost of a little expense when _writing_ it, although it's a very tough sell since it'd involve significant modifications to lots of codebases 

### named-parameters-mapping.js
should be recommended: no (very weak)
while it'd entail small modifications to existing codebases to make explicit, I
don't see why anyone would want anonymous mapping keys and values
we should, however, have something like the inverse of BaseDeprecation to
disable this rule on contracts with a compiler version pragma lower than 0.8.18

### named-return-values.js
should be recommended: yes (weak)
Naming a value always adds useful semantic information, and I can't really think
of a place where it's desireable for it to stay anonymous. However, we should
research if doing so adds the tracking for that variable to the stack, making
stack-too-deep errors more likely. If that's the case it should not be recommended.

### private-vars-leading-underscore.js
should be recommended: no (currently isn't, weak)

while it is a best practise, it's not widely used. And some projects use the
underscore to tell apart variables in state vs memory. Perhaps it'd be good to
have a different, non-recommended rule enforcing _that_?

### use-forbidden-name.js
should be recommended: yes (already is)

### var-name-mixedcase.js
should be recommended: yes (already is)
I believe it should actually be camelcase, since those starting with a capital
letter should be types, and those having underscores should be
constant/immutable but I don't know how to go about it.
Perhaps creating a var-name-camelcase rule, making that one recommended, and
de-recommending this one?

### func-order.js
should be deleted in favour of ordering

### imports-on-top.js
should be recommended:  yes (already is)

### ordering.js
should be recommended: yes (currently isn't)
Most projects do something similar to this, and I find those that don't usually
a bit harder to follow

### visibility-modifier-order.js
should be recommended: yes (already is) 

### avoid-call-value.js
should be recommended:  yes (already is)
I don't know if the compiler still considers this valid though
Docs could be improved.

### avoid-low-level-calls.js
should be recommended:  yes (already is)

### avoid-sha3.js
should be recommended: yes(already is)

### avoid-suicide.js
should be recommended: yes(already is)

### avoid-throw.js
should be recommended: yes(already is)

### avoid-tx-origin.js
should be recommended: yes(already is)

### check-send-result.js
should be recommended: yes (already is) 

### compiler-version.js
should be recommended:  yes (already is) 

### func-visibility.js
should be recommended:  yes (already is) 

### mark-callable-contracts.js
should be recommended: no (weak)
it's a tough sell but I'd like it ngl

### multiple-sends.js
should be recommended:  yes (already is)

### no-complex-fallback.js
should be recommended:  yes (already is)

### no-inline-assembly.js
should be recommended: yes (already is)

### not-rely-on-block-hash.js
should be recommended: yes (already is)

### not-rely-on-time.js
should be recommended: yes (already is, weak)
This is the rule I've seen manually disabled via an inline comment the most.
Could consider doing away with it, although it's true having the rule kind of
pushes all usages of block.timestamp to a single place a little
docs could be improved though.

### reentrancy.js
should be recommended: yes (already is)
docs could be improved

### state-visibility.js
should be recommended: yes (already is) 
docs could be improved
